### ISSUE TITLE
Discussion about abandoned registrations with no contact details counting towards the registration limit

### ISSUE THREAD

**Issue by lorenzocaum** [eventespresso/event-espresso-core#3324]
I found this comment (https://github.com/eventespresso/event-espresso-core/issues/732#issuecomment-430009452) about the system: 

>Registrations for abandoned registrations do count towards the registration limit of an event for as long as their session is active. This prevents events from over selling while someone is in the process of registering. As soon as the session expires however, any reserved tickets for the transaction are released and those spaces become available again.

One way that event organizers are adapting during the pandemic is by hosting smaller events with limited capacity.

A customer on Event Smart (https://secure.helpscout.net/conversation/1432346840/821694?folderId=284427) reached out to describe an issue that happens where there are a couple of spots left and a customer chooses more than is available.

This is not specific to Event Smart, and here is the structure of their event.

- Datetime for church service with a limit of 30

- Adult registration
- Child registration
- Another child registration option

Things work well until there are a few spots left. Here is an example.

Let's say there are 28 approved registrations which leaves 2 spots remaining. A mother is wants to register her family of 5.

She chooses 2 adult registrations and 3 child registrations and clicks register now.

She'll be taken to registration checkout where she sees a message about some spots not being available. Here is an example:

![Registration Checkout – Lorenzo Caum Site 2021-02-28 at 4 34 45 PM](https://user-images.githubusercontent.com/1079138/109434368-fc749280-79e2-11eb-9b93-5a66adc97549.jpg)

The message suggests a couple of ways forward: Cancel the registration and choose again or continue with the registration.

If the registration is finished, then they'll get some of the spots that they wanted. I think this approach makes sense around limited remaining tickets.

I didn't see a way to cancel the registration from registration checkout, and I spotted something interesting.

In testing, I noticed that fewer spots are available, or it can be sold out if I go back to the event page. 

For the customer on Event Smart, the event becomes temporarily sold out. This is because the reservation system holds spots as soon as someone selects some tickets / registration options and clicks Register Now. And if they click back as I did, then the spots that they previously selected are being held, but they can't immediately access them.

Someone could unknowingly take away available spots from an event with a couple of spots left by starting a registration and immediately abandoning.

Are there any drawbacks to having the reservation system run after someone completes step 1 of registration checkout? 

In other words, wait until there is at least a name and email, then hold the tickets in the reservation system?

**Comment by tn3rb** [comment]
> Someone could unknowingly take away available spots from an event with a couple of spots left by starting a registration and immediately abandoning.

yes this is the unfortunate downside to the current setup because it is extremely difficult to know when the registrtion process has been abandoned other than to set a timer and wait for it to expire. Right now that is essentially the session expiration which defaults to 60 minutes. 

While writing this response i had an idea that could potentially be a better solution for this problem. It may be possible to add some javascript to the registration checkout page that detects inactivity due to the lack of input on the page. After a set amount of inactivity, maybe 5 minutes, we could display a warning notice on the screen with a countdown timer giving them another 5 minutes to resume filling out the reg form. If that timer expires, then we cancel the registrations and release any spaces that were reserved for them. This of course only detects registration forms that sit idle and doesn't help with people that simply leave the page and navigate elsewhere. Although for that last issue we could also detect an attempt to leave the page and display a confirmation box requiring them to click OK or cancel (similar to when you try to leave the WP Post Editor after making changes that have not been saved). This would allow us to trigger an AJAX request to release their reservations.

> Are there any drawbacks to having the reservation system run after someone completes step 1 of registration checkout?

Yes, the downside to delaying reservations is that it allows more ppl to start into the registration process thinking that spaces are still available. In other words it allows people to successfully add tickets to their cart and proceed to the registration page where they can begin filling out the registration form, which for some sites is **extremely** long. Couple that with high volume traffic and things could get really messy. 

Consider the following scenario:

- company sends out an email to thousands of their employees about an internally organized event with limitied spaces availalbe

- the event is highly desireable to the employees who flood the site to register

- because reservations are deferred until later on, employees are able to select tickets and proceed to registration without any worries because the event is not sold out until payment is received

- unfortunately the registraion form is extremely complex and requires up to an hour to complete depending on the user. 

- in the time it takes enough people to fill out the registration form and reserve all of the available spaces, hundreds of employees have been able to select tickets and begin the hour long process of filling out the registration form, only to be told that spaces are no longer available. 

none of the above points is unheard of with Event Espresso customers, but together it can make for a "perfect storm" and result in a LOT of unhappy registrants.

**Comment by tn3rb** [comment]
@lorenzocaum 
any further comments on this discussion or can I close this?

**Comment by lorenzocaum** [comment]
Hi Brent, thanks for the reminder about this issue!

Here is more feedback that we've received on Event Smart. (Both are long time customers.)

Donna shared:

>I have an event that states sold out but out of 50 cap only 24 have registered but there is 26 under Rsrved column? What are these and how do I find what they are?

Clifford shared:

>The problem this creates is that those Rsrvd tickets now count towards the total, so the system shows the Event is SOLD OUT before we actually hit the 100 ticket limit. This required us to manually enter the system and start deleting Reserved tickets so that people could try again to buy. This is extremely difficult as it means someone has to stay logged into the system the whole time tickets are being sold to clean up the Rsrvd tickets constantly. Why is this happening, how can we get around this in the future?

I like the reservation system as I see how it prevents overselling. I also know we have a distributed product that needs to cater to different events. From my experience on Event Smart, most customers have registration forms that can be completed in less than 5 minutes, so I think the 1 hour period is plenty of time.

The challenge is it confuses customers when a spot is held even though no details have been collected.

@Pebblo Can you please review what I shared here (https://github.com/eventespresso/event-espresso-core/issues/3324#issue-818326116) and share your experience with Event Espresso?

**Comment by tn3rb** [comment]
Unfortunately there is no way to remove the reservation logic without going back to having overselling since that allows more people to begin the reg process than there are remaining spaces.

So are these people must be seeing their events sell out fast then if, in the case of Donna's example, 26 out of 50 are reserved, meaning that those 26 registrations happened within the hour, no? Were those registrations still reserved after an hour had passed? If so, then that is definitely an issue.

There's two potential solutions to this issue that I can think of:

 1. add an admin option for controlling the session length. This would allow admins to reduce their session time down to say a minimum of 15 minutes (probably should not go any lower than that) which can only be done now using a filter.

 2. reimplement the countdown timer on the checkout page. There used to be one but it could only be turned on via a filter and I'm not even sure it works anymore as I haven't touched it in like 8-9+ years. But basically we could run some javascript on the checkout page that only allows a certain amount of time to complete the registration process and then cancels the registration if time runs out. This is similar to the above but gives some direct feedback to the user. It would obviously be more complex to implement than the above but would likely be a more pleasant experience. It would also require some admin controls for setting the allowed registration time length.

**Comment by lorenzocaum** [comment]
Hi Brent, I want to keep the reservation system. I'm asking is its possible to delay when it starts until someone submits the registration form.

The examples that I shared earlier are the most common questions that we see. 

- A customer like Wendy may add custom questions. Then they want to check that it's working by starting a registration and they might go back and forth.
- A customer like Clifford may have popular events that are in high demand.

The confusion that we try to address is why the system reserves a spot when it has no details to follow up on. In other words, abandoned registrations with no attendee information hold spots that would otherwise be available to someone that is ready to promptly register and pay.

I recently helped a friend find a rare house plant on a website called Etsy. I saw that a seller had it and under the listing it showed something like "its in 3 people's carts." That may be true but my friend got the plant because she was the first to pay.

I think customers on Event Smart would be better off if the reservation ran when the registration form was submitted. 

1) An event organizer could see the attendee registrations that are temporarily holding spots.

2) The system could follow up through the pending payment email.

Can we keep the reservation system and introduce a filter that lets us change the timing of the reservation system to run a submit of the registration form?

@Pebblo Can you please check out the notes here (https://github.com/eventespresso/event-espresso-core/issues/3324#issue-818326116) and share your experience with Event Espresso?

**Comment by Pebblo** [comment]
> I recently helped a friend find a rare house plant on a website called Etsy. I saw that a seller had it and under the listing it showed something like "its in 3 people's carts." That may be true but my friend got the plant because she was the first to pay.

The use case for this on Etsy is a little deceiving because items are never automatically removed from your cart in Etsy, people leave items in their cart for weeks/months or longer ([Quick example](https://community.etsy.com/t5/Managing-Your-Shop/Item-in-someones-cart-How-do-you-know/m-p/134435184/highlight/true#M1277235)) 

I've also previously read about those being just completely false (not on Etsy, iirc it was some WooCommerce add-on) and just some JS that randomly makes up a value to make the item seem in demand,.... anyway I digress.

---

So the issue with events showing up as sold out due to reserved tickets does come up on the forums from time to time but not all that much really, when it does I usually recommend they use the current filter to switch the session lifespan to something shorter using this:

https://gist.github.com/Pebblo/b6e20b03c57efe693bd86ebcf1b00ce9

I recommend going no lower than 15mins when people ask for this and whilst I do understand the reasoning for the current default I think for the majority of users 15 minutes works fine. However, also note that the current system is hooking into one of our WP_CRON schedules which runs.... every hour, so just a shorter lifespan doesn't always mean shorter reserved times and we'd need another schedule for this but again it relies on WP_CRON so not always accurate.

An admin option for this would be good, but to work properly would also need some of what Brent suggested with the additional JS etc.

A 'better' reservation system would be great, something like the examples given below in some form would be awesome (I'm aware there is a lot to doing that) but I'm not so sure moving the reservation to submission of attendee info is the right move, I suspect it would take a fair bit of refactoring to do this and might just move the issue to another point in the process.

> I think customers on Event Smart would be better off if the reservation ran when the registration form was submitted.
> 1.  An event organizer could see the attendee registrations that are temporarily holding spots.
> 2.  The system could follow up through the pending payment email.

These are nice features but filtering registrations that are currently reserving spaces isn't available as is so would need additional work and we had follow up emails for offsite payments methods a while (which turned out to be a marmite feature (some loved some hated with no in between)), or would these be manual pending payment emails?

---

Looking at how the 'Big Boys' do this when NOT on a distributed system:

## Ticket master 
Give you an initial reserve time of 2mins when you select tickets, when you confirm you get another 2mins 30  to select Delivery options etc: https://monosnap.com/file/ZVb9SXKv3LwZELaKicZItruLeOWq0q

So they refresh the reservation time each time you submit 'something' by the looks of it.

At both of those points they must be reserving the actual tickets otherwise, the message they show when you time out doesn't really make sense:
https://monosnap.com/file/0fu6UssJ2DiGo9koVr3fP4qjTzfCH8
(The next option after the above if I'd have continued is to log in and pay)

## Skiddle
These guys give you a 6 min session time from when you submit your ticket selection. 

Selecting tickets show they were served for 6 mins and thats for a fixed form, Full name, email, confirm email, DOB, address, card details (plus from insurance options, donation option and 'lets stay in touch options')

https://monosnap.com/file/1TOZxEEbQCLlasNpGNCaKjpA8Jvv6o

So to me, reserving tickets when those tickets are selected makes sense and IF we do end up with a seating chart/allocation add-on for EE4 I'm not sure how well that would work if reservations were made after the attendee info. You'd want to know you reserved your 2 tickets next to each other before submitting details etc.

**Comment by lorenzocaum** [comment]
Hey @Pebblo @tn3rb , thanks for sharing feedback. Based on the discussion, I'm going to talk to Nazar about testing a shorter session lifespan of 30 minutes on Event Smart.

Qubit did an analysis of 6700 ecommerce experiments and here is what worked:

• scarcity (stock pointers) +2.9% uplift
• **urgency (countdown timers) +2.3% uplift**
• social proof (informing users of others’ behaviour) +1.5% uplift
• abandonment recovery (messaging to keep users on-site) +1.1% uplift
• product recommendations (suggesting other products to purchase) +0.4% uplift

Most simple UI changes to websites are ineffective. For example

• colour (changing the colour of elements on a website) +0.0% uplift
• buttons (modifying website buttons) -0.2% uplift
• calls to action (changing the wording on a website to be more suggestive) -0.3% uplift

I can't find the PDF on their website and the notes above are from my bookmark.

The countdown timer would be good enhancement to add later on. It could nudge someone to finish their registration.

**Comment by tn3rb** [comment]
plz see #3871

**Comment by tn3rb** [comment]
closing because there is now an admin option that allows users to set the session lifespan without even having to use a filter
